README.md
![Spam Guard](resources/img/spamguard.png)

# Spam Guard 0.6.0

Lovingly crafted by [Selvin Ortiz][developer] for [Craft CMS][craftcms]

**Spam Guard** harnesses the power of [Akismet][akismet] to fight **Spam**

## Features
1. Proven _Spam Fighting_ capabilities with the help of [Akismet][akismet]
2. Native support for the [Contact Form Plugin][contactform]
3. Native support for the [Guest Entries Plugin][guestentries] 
4. Native support for the [Sprout Forms](http://sprout.barrelstrengthdesign.com/craft-plugins/forms)
5. Clean and efficient codebase with good documentation
6. Friendly developer support via **Github**, [Twitter][developer], and via email at <selvin@selv.in>

## Installation
1. Download the [Official Release][release]
2. Extract the archive into your `craft/plugins` directory
4. Install **Spam Guard** from the control panel **@** `Settings > Plugins`
5. Add your `API Key` and `Origin URL`
6. Enable support for [Contact Form][contactform] and/or [Guest Entries][guestentries]

## Upgrading
If you are upgrading to version **0.6.0** from an earlier version, please see the documentation for **Guest Entries** because the way hidden fields in your form are defined has changed.

1. Spam Guard now expects the following fields `spamguard[emailField|authorField|contentField]`
2. The values for those fields should now be wrapped in curly braces match entry field handles

## Dependencies
- PHP 5.3.2 _or above_
- Craft 1.3 Build 2415 _or above_
- [Contact Form 1.3+][contactform] _for contact form support_
- [Guest Entries 1.1+][guestentries]  _for guest entries support_

_Please note that [Guest Entries][guestentries] requires Craft **2.0** or above_

---

### Contact Form Usage
To fight spam coming from your contact form, you must do the following...

1. Follow the [Contact Form][contactform] setup guide if you haven't already
2. Make sure [Contact Form][contactform] support **@** `Settings > Plugins > Spam Guard`

_That is it, all future submissions will be monitored by spam guard_ **;)**

### Guest Entries Usage
To fight spam coming from your guest entry form, you must do the following...

1. Follow the [Guest Entries][guestentries] setup guide if you haven't already
2. Make sure [Guest Entries][guestentries] support is enabled **@** `Settings > Plugins > Spam Guard`
3. Add the `hidden input` fields to your form so **Spam Guard** knows what to validate

#### Hidden Input Fields
Add the following fields to your guest entry form.

```html
<input type="hidden" name="spamguard[emailField]" value="{guestEntryEmailFieldHandle}">
<input type="hidden" name="spamguard[authorField]" value="{guestEntryFullNameFieldHandle}">
<input type="hidden" name="spamguard[contentField]" value="{guestEntryBodyFieldHandle}">
```

These fields need to be defined so that **Spam Guard** knows what attributes to look for in the **guest entry** in order to prepare the data to pass along to [Akismet][akismet] for validation.

When the form is submitted and the **Guest Entry** is validated, the entry will be handed to **Spam Guard** which will then grab the `spamguard[emailField|authorField|contentField]` values containing **twig** placeholders which will then be **replaced** by attribute values found in the [EntryModel](http://buildwithcraft.com/docs/templating/entrymodel)

_Note that the emailField and authorField are not required!_

### Sprout Forms Usage
Sprout Forms support is implemented in the same way support is implemented for Guest Entries, the directions for Guest Entries should be followed. 

---

## Changelog
All noteworthy changes are listed on the official [changelog][changelog]

## Feedback
If you have any feedback or questions, please reach out to me on twitter [@selvinortiz][developer]


## Credits
* Huge thanks to the **WordPress/Akismet** team for their great efforts and service they provide.
* _Special thanks to [@themccallister](https://github.com/themccallister) for testing Spam Guard and for providing valuable feedback._

## License
Spam Guard is open source software licensed under the [MIT license][license]

![Open Source Initiative][osilogo]

[developer]:http://twitter.com/selvinortiz "@selvinortiz"
[release]:https://github.com/selvinortiz/craft.spamguard/releases "Official Release"
[license]:https://raw.github.com/selvinortiz/craft.spamguard/master/LICENSE "MIT License"
[changelog]:https://github.com/selvinortiz/craft.spamguard/blob/master/CHANGELOG.md "Changelog"
[craftcms]:http://buildwithcraft.com "Craft CMS"
[akismet]:http://akismet.com "Akismet"
[contactform]:https://github.com/pixelandtonic/ContactForm "Contact Form"
[guestentries]:https://github.com/pixelandtonic/GuestEntries "Guest Entries"
[pixelandtonic]:http://pixelandtonic.com "Pixel & Tonic"
[osilogo]:https://github.com/selvinortiz/craft.spamguard/raw/master/resources/img/osilogo.png "Open Source Initiative"


CHANGELOG.md
# Spam Guard

Lovingly crafted by [Selvin Ortiz][developer] for [Craft CMS][craftcms]

## Changelog

### 0.6.0
- Adds support for [Sprout Forms][sproutforms]
- Adds the `SpamGuardInvalidKeyException` class to the `Craft` namespace
- Improves the **Logged Submissions** template with faster load times
- Improves documentation for **Guest Entries** support
- Removes external libraries
- Removes composer dependency
- Removes the `SpamGuardVariable` class
- Removes the `spamGuardDetectSpam()` plugin hook
- Removes the `InvalidKeyException` class

### 0.5.5
- Adds `InvalidKeyException` reference to `submitSpam()` and `submitHam()`
- Adds the ability to use `{siteUrl}` as the **Origin URL** in settings
- Fixes missing `InvalidKeyException` reference in issue #7

### 0.5.4
- Fixes deprecation issue #6 in Craft 2.0
- Fixes some typos in code comments and the README
- Improves styling for logged submissions

### 0.5.3
- Adds native support for [Guest Entries][guestentries] by [P&T][pixelandtonic]
- Fixes a typo in the footer from code reuse outlined on issue #2
- Improves the way the *User IP* is retrieved
- Improves settings handling methods

### 0.5.2
- Adds automatic redirection to `Settings > Plugins > Spam Guard` after install

### 0.5.1
- Adds setting to opt out of submission logging
- Adds the audit column properties to the `SpamGuardModel`
- Fixes line endings and platform inconsistencies
- Improves the layout and clarity of the settings template

### 0.5.0
_This version introduces breaking changes, a fresh installed is required_

- Adds native support for [Contact Form][contactform] by [P&T][pixelandtonic]
- Adds an [Akismet][akismet] client built on **Guzzle**
- Adds a basic Unit Test Suite
- Adds basic submission logging
- Improves codebase by removing proprietary form functionality

### 0.4.7
- Fixes an issue caused by accessing an undefined constant @ `SpamGuardService`
- Improves the settings UI by adding icons and a setting for [Contact Form][contactform]
- Removes the bridge package that was used to autoload the **Akismet** library

### 0.4.6
- Removes class constants for plugin properties
- Adds `plugin.json` to define plugin properties
- Cleans up trailing spaces and enforces new line (EOF)

### 0.4.5
- Adds editor behavior to the `emailTemplate` textarea

### 0.4.4
- Removes `actionSpamGuardTest()`
- Adds `SpamGuard_MessageModel`
- Adds `SpamGuard_MessagingService`
- Adds `actionSendMessage()` to enable contact form submissions

### 0.4.3
- Renames `rocket` to `bridge`
- Makes the `bridge` helper library play nice with other plugins
- Adds `safeOutput()` to flag content as safe and avoid using the raw filter

### 0.4.2
_This version contains breaking changes meant to improve the foundation to build on_

- Adds `spamGuardDetectSpam()` to `SpamGuardPlugin` *LTS*
- Adds `detectSpam` to the `SpamGuardService` *LTS*
- Removes `spamGuardPostedContent()` from `SpamGuardPlugin`
- Removes `spamGuardSubmittedContent()` from `SpamGuardPlugin`
- Removes `isSpam` from `SpamGuardService`

### 0.4.1
- Adds the `Akismet` model
- Adds the `spamGuardSubmittedContent()` action
- Adds placeholder Controller/Actions to be implemented
- Fixes inaccurate fetching of setting values
- Improves missing `API Key` handling
- Removes code examples in source code
- Removes empty `hookRegisterCpRoutes()`

### 0.4.0
- Removes `composer.json` and the composer `vendor` package
- Improves the `SpamGuardService` constructor
- Removes `rocket/packages` from `.gitignore` since composer is not being used

### 0.3.0
- Adds the `spamGuardPostedContent()` action
- Adds callback support based on `spamGuardPostedContent()` results
- Improves semantics by placing complex logic in the `SpamGuardService`
- Improves `SpamGuardController` by only exposing the `actionIsSpam()`

### 0.2.0
- Adds `SpamGuardModel`
- Adds `SpamGuardController`
- Adds `Rocket::loadClass()` to load dependencies
- Removes composer dependency

### 0.1
Initial proof of concept release

[developer]:http://twitter.com/selvinortiz "@selvinortiz"
[craftcms]:http://buildwithcraft.com "Craft CMS"
[akismet]:http://akismet.com "Akismet"
[contactform]:https://github.com/pixelandtonic/ContactForm "Contact Form"
[guestentries]:https://github.com/pixelandtonic/GuestEntries "Guest Entries"
[sproutforms]:http://sprout.barrelstrengthdesign.com/craft-plugins/forms "Sprout Forms"
[pixelandtonic]:http://pixelandtonic.com "Pixel & Tonic"


library/Kismet.php
<?php
namespace selvinortiz\spamguard;

use Guzzle\Http\Client;

/**
 * Kismet is a simple and modern client for the Akismet API
 *
 * @author		Selvin Ortiz <selvin@selvin.co>
 * @package		spamguard
 * @version		1.0
 * @license		http://opensource.org/licenses/MIT Copyright 2014 Selvin Ortiz
 * @copyright	Selvin Ortiz
 */
class Kismet
{
	protected $apiKey;
	protected $originUrl;
	protected $httpClient;

	const ENDPOINT = 'rest.akismet.com/1.1';

	public function __construct(array $data=array())
	{
		$this->apiKey		= $this->fetch('apiKey', $data, null);
		$this->originUrl	= $this->fetch('originUrl', $data, null);
		$this->httpClient	= $this->fetch('httpClient', $data, new Client());
	}

	public function isKeyValid(array $data=array())
	{
		$params = array(
			'key'	=> $this->fetch('apiKey', $data, $this->apiKey),
			'blog'	=> $this->fetch('originUrl', $data, $this->originUrl)
		);

		$request	= $this->httpClient->post($this->getKeyEndpoint(), null, $params);
		$response	= (string) $request->send()->getBody();

		return (bool) ($response == 'valid');
	}

	public function isSpam(array $data=array())
	{
		$params = $this->mergeWithDefaultParams(
			array(
				'comment_author'		=> $this->fetch('author', $data),
				'comment_content'		=> $this->fetch('content', $data),
				'comment_author_email'	=> $this->fetch('email', $data)
			)
		);

		if ($this->isKeyValid())
		{
			$request	= $this->httpClient->post($this->getContentEndpoint(), null, $params);
			$response	= (string) $request->send()->getBody();

			return (bool) ('true' == $response);
		}

		throw new Exception('The key provided is not valid or has expired.');
	}

	public function submitSpam(array $data=array())
	{
		$params = $this->mergeWithDefaultParams(
			array(
				'comment_author'		=> $this->fetch('author', $data),
				'comment_content'		=> $this->fetch('content', $data),
				'comment_author_email'	=> $this->fetch('email', $data)
			)
		);

		if ($this->isKeyValid())
		{
			$request	= $this->httpClient->post($this->getSpamEndpoint(), null, $params);
			$response	= (string) $request->send()->getBody();

			return (bool) ('Thanks for making the web a better place.' == $response);
		}

		throw new Exception('The key provided is not valid or has expired.');
	}

	public function submitHam(array $data=array())
	{
		$params = $this->mergeWithDefaultParams(
			array(
				'comment_author'		=> $this->fetch('author', $data),
				'comment_content'		=> $this->fetch('content', $data),
				'comment_author_email'	=> $this->fetch('email', $data)
			)
		);

		if ($this->isKeyValid())
		{
			$request	= $this->httpClient->post($this->getHamEndpoint(), null, $params);
			$response	= (string) $request->send()->getBody();

			return (bool) ('Thanks for making the web a better place.' == $response);
		}

		throw new Exception('The key provided is not valid or has expired.');
	}

	protected function mergeWithDefaultParams(array $extraParams=array())
	{
		return array_merge(
			array(
				'blog'			=> \Craft\craft()->getSiteUrl(),
				'user_ip'		=> $_SERVER['REMOTE_ADDR'],
				'user_agent'	=> $this->getUserAgent(),
				'comment_type'	=> 'Entry'
			),
			$extraParams
		);
	}

	protected function getUserAgent()
	{
		return 'Craft 1.3 | Spam Guard 0.5';
	}

	protected function fetch($key, array $arr=array(), $def=false)
	{
		return array_key_exists($key, $arr) ? $arr[$key] : $def;
	}
	
	protected function getKeyEndpoint()
	{
		return sprintf('http://%s/verify-key', self::ENDPOINT);
	}

	protected function getContentEndpoint()
	{
		return sprintf('http://%s.%s/comment-check', $this->apiKey, self::ENDPOINT);
	}

	protected function getSpamEndpoint()
	{
		return sprintf('http://%s.%s/submit-spam', $this->apiKey, self::ENDPOINT);
	}

	protected function getHamEndpoint()
	{
		return sprintf('http://%s.%s/submit-ham', $this->apiKey, self::ENDPOINT);
	}
}
